### PR TITLE
wolfSSL as a Zephyr module (without setup.sh)

### PR DESCRIPTION
**Pull request by maximevince**
Cleaner way to implement wolfSSL as a Zephyr module
* No setup.sh script has to be run by the user,
* No files are added or copies to zephyr directories which are not tracked by git
* Following Zephyr / west's module guidelines:  
https://docs.zephyrproject.org/latest/guides/modules.html#build-system-integration and https://docs.zephyrproject.org/latest/guides/modules.html#using-west

There would just be one step required for the user: modifying the west.yml and running `west update`
(just like your script is requesting to do)

E.g.:
```yaml
manifest:
  remotes:
    # <your other remotes>
    - name: wolfssl
      url-base: https://github.com/maximevince # can become 'github.com/wolfssl' after merging

  projects:
    # <your other projects>
    - name: wolfssl
      path: modules/crypto/wolfssl
      revision: zephyr-module-support # can become 'master' after merging
      remote: wolfssl
```

That's it.

Now west recognizes 'wolfssl' as a module, and will include it's Kconfig and CMakeFiles.txt in the build system:
```sh
% west list | grep wolfssl
wolfssl      modules/crypto/wolfssl       zephyr-module-support                    https://github.com/maximevince/wolfssl
```  
<br>

NOTE: I have temporarily commented out the `include IDE/zephyr/include.am` from `IDE/include.am` for now.
I don't know exactly how this should be handled within wolfSSL. Let me know what to do here. (7532ac530a4f34ef89b4056ef773ab81684527ef)

### CODE REVIEW

**Comment by SparkiDev** [comment]
Worked for me!
Thanks @maximevince.

**Comment by SparkiDev** [comment]
ok to test

**Review by SparkiDev — APPROVED** [review]

**Comment by maximevince** [comment]
A few option points:
* How to handle the commented out include `IDE/zephyr/include.am` from `IDE/include.am` ? Does anything have to be done about this?
* It's not really easy to use a custom `user_settings.h` file right now, except if I don't understand how to do it correctly.  
 I tried using the `CONFIG_WOLFSSL_SETTINGS_FILE="user_settings-tls13.h` in my prj.conf, but it can't find the path to my include file while building the wolfSSL module, since the module's include paths do not include my application's include paths.
* I had to manually add WOLFSSL definitions to my application's CMakeLists.txt, to match those of the wolfSSL module. E.g.:   
  ```
  add_definitions("-DWOLFSSL_ZEPHYR")
  add_definitions("-DNO_FILESYSTEM")
  add_definitions("-DWOLFSSL_USER_SETTINGS")
  ```
  I am not sure that's what we want, but it works for now. Does anybody know the right way to do that? Perhaps there's a possibility in the woflSSL module CMakeLists.txt to indicate that these definitions should be used by the application linking this library as well?

**Comment by dgarske** [comment]
Hi @maximevince , thanks for being one of our approved contributors and this excellent PR! Sean is away on vacation, so I will help drive this PR to completion. Our automated Jenkins testing is reporting an issue with new files not added to the include.am.

Report:
```
-----------------------------------------------
Detecting new files that have been added
-----------------------------------------------
added: zephyr/CMakeLists.txt
added: zephyr/README.md
added: zephyr/module.yml
added: zephyr/wolfssl/options.h
Total new files to check: 4
-----------------------------------------------
Checking if files are included in an include.am
-----------------------------------------------
./zephyr/include.am:EXTRA_DIST+= IDE/zephyr/README.md
./zephyr/include.am:EXTRA_DIST+= IDE/zephyr/lib/zephyr/module.yml
AUTOCONF NEW FILE DETECTOR FAILED!
```

Can you make sure these news files are added to include.am? Also I noticed `include IDE/zephyr/include.am` is commented out? These can be `EXTRA_DIST += ` so our `make dist` will include these.

Thanks,
David Garske, wolfSSL

**Comment by maximevince** [comment]
> Hi @maximevince , thanks for being one of our approved contributors and this excellent PR! Sean is away on vacation, so I will help drive this PR to completion. Our automated Jenkins testing is reporting an issue with new files not added to the include.am.
> 
> Report:
> 
> ```
> -----------------------------------------------
> Detecting new files that have been added
> -----------------------------------------------
> added: zephyr/CMakeLists.txt
> added: zephyr/README.md
> added: zephyr/module.yml
> added: zephyr/wolfssl/options.h
> Total new files to check: 4
> -----------------------------------------------
> Checking if files are included in an include.am
> -----------------------------------------------
> ./zephyr/include.am:EXTRA_DIST+= IDE/zephyr/README.md
> ./zephyr/include.am:EXTRA_DIST+= IDE/zephyr/lib/zephyr/module.yml
> AUTOCONF NEW FILE DETECTOR FAILED!
> ```
> 
> Can you make sure these news files are added to include.am? Also I noticed `include IDE/zephyr/include.am` is commented out? These can be `EXTRA_DIST += ` so our `make dist` will include these.
> 
> Thanks,
> David Garske, wolfSSL

Okay, done. Let's see if the test passes now.

**Review by dgarske — CHANGES_REQUESTED** [review]
Hi @maximevince ,

Great work on this. I was able to setup zephyr and test out having wolfSSL as a module. All worked great. Just one note about the include location.

Also is it really required to have everything in `<wolfssl-root>/zephyr`? My preference is in `IDE/zephyr`, but if that is not possible I think we can accept. Just wanted to confirm.

For the build options, I like `WOLFSSL_USER_SETTINGS` and user_settings.h. Are you seeing any issues with that?

Thanks,
David Garske, wolfSSL

File: IDE/include.am [hunk 1]
```diff
@@ -38,7 +38,7 @@ include IDE/XilinxSDK/include.am
 include IDE/VisualDSP/include.am
 include IDE/QNX/include.am
 include IDE/WINCE/include.am
-include IDE/zephyr/include.am
+include zephyr/include.am
```
[the remark this answers is not in the document]
Comment (maximevince) [reply IDE/include.am:41]: Ok, sure!

**Comment by maximevince** [comment]
> Hi @maximevince ,
> 
> Great work on this. I was able to setup zephyr and test out having wolfSSL as a module. All worked great. Just one note about the include location.
> 
> Also is it really required to have everything in `<wolfssl-root>/zephyr`? My preference is in `IDE/zephyr`, but if that is not possible I think we can accept. Just wanted to confirm.
> 
> For the build options, I like `WOLFSSL_USER_SETTINGS` and user_settings.h. Are you seeing any issues with that?
> 
> Thanks,
> David Garske, wolfSSL

Alright, good to hear!

I am afraid it has to be in `<wolfssl-root>/zephyr`, since that's the way the `west` tool finds the modules.
From https://docs.zephyrproject.org/latest/guides/modules.html#using-west:
> * If the project contains a file named `zephyr/module.yml`, then the content of that file will be used to determine which files should be added to the build, as described in the previous section.
> * Otherwise (i.e. if the project has no `zephyr/module.yml`), the build system looks for `zephyr/CMakeLists.txt` and `zephyr/Kconfig` files in the project. If both are present, the project is considered a module, and those files will be added to the build.
> * If neither of those checks succeed, the project is not considered a module, and is not added to **ZEPHYR_MODULES**.

If you prefer, we can move all the files in the `<wolfssl-root>/zephyr` back to `<wolfssl-root>/IDE/zephyr`, except for `module.yml`, which has to stay at `<wolfssl-root>/zephyr/module.yml` ?


For the build options: Yes, in principle `WOLFSSL_USER_SETTINGS` is fine, but the problem is that the way it is now, this has to point to an absolute path. This is because Zephyr does not add the application's include paths when building a module/library (which wolfSSL now is). You'd rather not have an application change files in the module's directories, so that `user_settings.h` file will probably reside in the application's directory. There is probably a way to do this in CMake/west/Zephyr, but I have yet to find it.

**Review by dgarske — APPROVED** [review]
All looks great and tests out. I am okay with keeping the zephyr directory in the root. Very nice having wolfSSL setup as a Zephyr module. Great work on this Sean and Max. Once Jenkins is done testing I will merge.

**Comment by maximevince** [comment]
Cool!

**Pull request merged**